File: 프로그래머스/3/42895. N으로 표현/N으로 표현.cpp
#include <bits/stdc++.h>

using namespace std;
int target;
int n;
int ans = 9;

void dfs(int cur, int cnt){
    if(cur == target){
        ans = min(ans,cnt);
    }
    int t = n;
    for(int i=0;i<8-cnt;i++){
        dfs(cur+t,cnt+i+1);
        dfs(cur-t,cnt+i+1);
        dfs(cur*t,cnt+i+1);
        dfs(cur/t,cnt+i+1);
        t = n + t*10;
    }
}
int solution(int N, int number) {
    target = number;
    n = N;
    dfs(0,0);
    return ans == 9 ? -1 : ans;
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;
class Solution {
    int target;
    int ans;
    int n;
    public void dfs(int cur, int cnt){
        if(cur==target){
            ans = Math.min(cnt,ans);
        }
        int tmp = n;
        for(int i=1;i<=8-cnt;i++){
            dfs(cur+tmp,cnt+i);
            dfs(cur-tmp,cnt+i);
            dfs(cur/tmp,cnt+i);
            dfs(cur*tmp,cnt+i);
            tmp = n + tmp*10;
        }
    }
    public int solution(int N, int number) {
        n = N;
        target = number;
        ans = 10;
        dfs(0,0);
        if(ans==10){
            return -1;
        }
        return ans;
    }
}